### PR TITLE
fix(python): Fix interchange protocol for new String type

### PR DESCRIPTION
**Pull request by stinodego**
Ref #13763

#### Changes

* Rewrite `_get_buffer` to `_get_buffers`, returning all buffers at the same time.
* Implement `_get_buffers` for new String type
* Update the interchange protocol implementation to raise `CopyNotAllowedError` when converting to/from String types. This also affects Categoricals as the categories are a String column.
* Update tests to match
* Update tests for fixes in new pandas version (2.2.0)

**IMPORTANT:** We must still fix the `Enum` dtype to be zero-copy on conversion from Rust to Python, otherwise we cannot guarantee zero copy at all.

### CODE REVIEW

**Comment by ritchie46** [comment]
> IMPORTANT: We must still fix the Enum dtype to be zero-copy on conversion from Rust to Python, otherwise we cannot guarantee zero copy at all.

Nice one @stinodego. I shall do the `Enum` type and then we can issue a pre-release. :)

File: py-polars/src/series/buffers.rs [hunk 1]
```diff
@@ -104,85 +123,26 @@ fn get_bitmap(s: &Series) -> Option<PySeries> {
...
-            }
-            return Ok(Series::from_vec(s.name(), offsets).into());
-        },
-        _ => return Err(PyValueError::new_err("expected list/string")),
-    };
-    let buffers = buffers
-        .map(|arr| PrimitiveArray::from_data_default(arr.buffer().clone(), None).boxed())
-        .collect::<Vec<_>>();
-    Ok(Series::try_from((s.name(), buffers))
+fn get_string_bytes(arr: &Utf8Array<i64>) -> PyResult<PySeries> {
```
Comment (ritchie46) [inline py-polars/src/series/buffers.rs:126]: Maybe a add a comment that we don't return the buffers in the `Series`, but allocate buffers that match the expectations of the protocol.

File: py-polars/polars/interchange/column.py [hunk 2]
```diff
@@ -148,32 +150,46 @@ def get_chunks(self, n_chunks: int | None = None) -> Iterator[PolarsColumn]:
 
     def get_buffers(self) -> ColumnBuffers:
         """Return a dictionary containing the underlying buffers."""
```
Comment (ritchie46) [inline py-polars/polars/interchange/column.py:152]: I think this docstring must be adapted to clearly explain that for `String` and  `Binary` types we return a `buffer` that matches expectations of the DataFrame protocol.

Maybe this function should be private as well? Then we can change the outputs more easily if we have something better.

File: py-polars/src/series/buffers.rs (hunk 1 above)
Comment (stinodego) [reply py-polars/src/series/buffers.rs:126]: I have some documentation on this in the Python docstring, but I'll do some more documentation here.

File: py-polars/polars/interchange/column.py (hunk 2 above)
Comment (stinodego) [reply py-polars/polars/interchange/column.py:152]: This is the protocol itself, we are not allowed to change it. In the Series method `Series._get_buffers` I have included this information:

```
        Warnings
        --------
        The underlying buffers for `String` Series can not be represented in this
        format. Instead, the buffers are converted to a values and offsets buffer.
        This is used for the dataframe interchange protocol.
```

**Comment by stinodego** [comment]
I added additional documentation that hopefully makes it abundantly clear that buffer operations on String Series require conversion from/to offsets and bytes buffers.

File: py-polars/polars/interchange/column.py (hunk 2 above)
Comment (ritchie46) [reply py-polars/polars/interchange/column.py:152]: Oh, right. I didn't read properly. As long as it is not public `get_buffer` on our core datastructures it is fine.

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**